Integrate OuinetService-related features into Ouinet AAR

paidforby opened the pull request:
This PR introduces a new option for managing the ouinet client as a service that can be kept alive while the application is in background by creating a foreground notification. 

It is largely a generalization of code that already existed in the Ceno Browser, but makes it customizable and reusable for other implementations of ouinet. This resolves an open issue against the Ceno Browser, https://gitlab.com/censorship-no/ceno-browser/-/issues/39, to remove Ouinet-specific code from Ceno and move it into the ouinet AAR.

To see an example of how these new features are to be used see [these lines](https://github.com/paidforby/ouinet-examples/blob/android-ouinet-service/android/kotlin/app/src/main/java/ie/equalit/ouinet_examples/android_kotlin/MainActivity.kt#L42-L79) in my fork of the ouinet-examples repo.

Also consider this PR a place to discuss possibly organizing the new features differently. As it stands, I've left all the classes in the root `ie.equalit.ouinet` package. I couldn't really see a good way of grouping the new classes and don't want to move the old classes, since that could break backwards compatibility. 

Maybe leaving everything in the root packages makes sense? Or maybe there is some way in which we would like to structure the package by adding more sub-packages (as additional directories)? Thoughts?

paidforby committed:
Get ouinetVersionName from from version.txt and get ouinetBuildId with git-version-string.gradle instead of shell script, related to #58

paidforby committed:
Add kotlin classes for integration of OuinetService in android library.

Includes notification builder, a background class that manages
the OuinetService, and default resources for the notification UI.
Also adds kotlin-android plugins to build.gradle for supporting
compilation of kotlin source code.

mhqz commented:
Thank you @paidforby for this effort that will simplify a lot future integrations :metal: 

Regarding to the changes applied to the support library dependencies (support-annotations, etc) and the new minimum `OUINET_TARGET_API`; could you please update the related docs? [doc/android-sdk-versions.md](https://github.com/equalitie/ouinet/blob/master/doc/android-sdk-versions.md)

mhqz commented:
> Maybe leaving everything in the root packages makes sense? Or maybe there is some way in which we would like to structure the package by adding more sub-packages (as additional directories)? Thoughts?

For this particular PR I'd suggest to keep the new classes in the root (I usually prefer a flat organization strategy that allows you to use the whole library than a modular one that applies only for a certain parts of the code).

We can open a separate ticket to discuss a new package structure not only for the Android libraries but for the C++ components so we can expose solid bindings (for other languages) and reuse them in other projects.

mhqz commented on android/build-scripts/git-version-string.gradle:
```diff
@@ -0,0 +1,22 @@
+ext.gitVersionString = { ->
+    def versionString = gitBranch() + " " + gitId()
+    return versionString
+}
```
Missing newline at the end of the file.

mhqz commented on android/ouinet/src/main/java/ie/equalit/ouinet/ConnectivityBroadcastReceiver.kt:
```diff
@@ -0,0 +1,47 @@
+        val state = info.state
+        if (state == NetworkInfo.State.CONNECTED || state == NetworkInfo.State.DISCONNECTED) {
+            if (context != null) {
+                //logger.debug("Network state changed to ${state}")
```
Is this logger still needed?

paidforby replied:
Nope, I think it was just a holder over from porting this from the Ceno code base, I'll remove it.

mhqz commented on android/ouinet/src/main/java/ie/equalit/ouinet/ConnectivityBroadcastReceiver.kt:
```diff
@@ -0,0 +1,47 @@
+        val info = intent.getParcelableExtra<NetworkInfo>(ConnectivityManager.EXTRA_NETWORK_INFO)
+            ?: return
+
+        // Restart the Ouinet client whenever connectivity has changed and become stable.
```
This comment is useful but I think code block below could be encapsulated in a method, e.g.  `restartOnConnectivityChange(..)`

paidforby replied:
Good suggestion! If something needs a comment, just put in a method :slightly_smiling_face:. I didn't refactor this ConnectivityBroadcastReceiver code very much because it has a ton of deprecation warnings, so it might need a complete overhaul sooner rather than later.

mhqz commented on android/build-scripts/git-version-string.gradle:
```diff
@@ -0,0 +1,22 @@
+def gitBranch() {
```
Great idea to move this logic to Gradle scripts! :metal::smile_cat:

mhqz commented on android/ouinet/src/main/java/ie/equalit/ouinet/OuinetNotification.kt:
```diff
@@ -0,0 +1,155 @@
+                .addAction(config.clearIcon, config.confirmText, confirmPIntent)
+            handler.postDelayed(
+                showConfirmCallback,
+                3000 /* ms */
```
Not sure if this is a good idea but wanted to share it with you anyway :smile_cat: 

What about replacing the comment `/* ms */` with a constant?
e.g. `3 * MILLISECONDS`

paidforby replied:
I think it's a great idea. This `/* ms */` was a holder over from the original code from Ceno v1.6.x code, so I have no attachment to it, I'll replace with a constant.

mhqz reviewed (commented):
The new kotlin code for the OuinetService looks great, thank you @paidforby :slightly_smiling_face: 

I've added some comments about minor fixes and improvements to keep the code clean.

Please let me know if you have any questions.

paidforby committed:
Incorporate suggestions from @mhqz for #60,

 - New line at end of git-version-string.gradle
 - Put restart of Ouinet on connectivity change into it's own method
 - Add MILLISECOND constant for notification interval and delay

paidforby commented:
Thanks for the suggestions @mhqz. I've made the recommended changes in a new commit. Please take a look to make sure I didn't misinterpret any of your notes. If it all looks good, I think we are good to merge and release this feature in `v0.21.7` (unless you think it justifies going to `v0.22.0`?)

paidforby committed:
Add gradle.properties file with android.useAndroidX set and a few other values to allow build to succeed, related to #60

mhqz commented:
[567df30](https://github.com/equalitie/ouinet/pull/60/commits/567df3063cbed05d962969c4b703489419f08a99) looks fine to me, thanks a lot for applying the changes :+1:

I just didn't see any commit related to the update of the SDK docs: https://github.com/equalitie/ouinet/pull/60#issuecomment-1448932918
